apollo-federation/src/operation/merging.rs

```diff
@@ -173,116 +175,222 @@ impl SelectionSet {
-                        let Selection::InlineFragment(other_inline_fragment_selection) =
-                            other_selection
+                    Selection::InlineFragment(inline) => {
+                        let MatchedSelectionBuffer::InlineFragment(buffer) = entry
+                            .or_insert_with(|| MatchedSelectionBuffer::InlineFragment(Vec::new()))
                         else {
                             return Err(FederationError::internal(
-                                format!(
-                                    "Inline fragment selection key under parent type \"{}\" {}references non-field selection",
-                                    self_inline_fragment_selection.inline_fragment.parent_type_position,
-                                    self_inline_fragment_selection.inline_fragment.type_condition_position.clone()
-                                        .map_or_else(
-                                            String::new,
-                                            |cond| format!("(type condition: {}) ", cond),
-                                        ),
+                            format!(
+                                "Inline fragment selection key under parent type \"{}\" {}references non-field selection",
+                                inline.inline_fragment.parent_type_position,
+                                inline.inline_fragment.type_condition_position.clone()
+                                    .map_or_else(
+                                        String::new,
+                                        |cond| format!("(type condition: {}) ", cond),
+                                    ),
                                 ),
                             ));
                         };
-                        inline_fragments
-                            .entry(other_key)
-                            .or_insert_with(Vec::new)
-                            .push(other_inline_fragment_selection);
+                        println!("Buffering an inline fragment: {inline}");
+                        buffer.push(inline);
+                    }
+                }
+                Ok(())
+            }
+        }
+
+        // We need to maintain the order of the inner selection buffers.
+        let mut buffer = SelectionBuffer(IndexMap::default());
+
+        // This is a minor optimization. We could check this in the "recurse" function, but it will
+        // never change.
+        if try_flattening_fragments {
+            fn recurse_on_inline_fragment<'a>(
+                buffer: &mut SelectionBuffer<'a>,
+                type_pos: &CompositeTypeDefinitionPosition,
+                others: impl Iterator<Item = &'a Selection>,
+            ) -> Result<(), FederationError> {
+                for other_selection in others {
+                    if let Selection::InlineFragment(inline) = other_selection {
+                        if inline.inline_fragment.data().directives.len() != 0
```

Comment:
I think this is problematic -> we can only flatten the fragment IFF parent type == type condition AND (either we don't have any directives OR we have same directives as the parent)

If there is no explicit type condition we still need to compare directives.